### ISSUE TITLE
Do we need `attribute` ,`notifications_count` parameters?

### ISSUE THREAD

**Issue by artificis** [jonhue/notifications-rails#2]
https://github.com/jonhue/notifications-rails/blob/f160b35446836171f6940b1c1bdbfd304f153694/notification-renderer/app/helpers/notification_renderer_helper.rb#L3

I'm not sure what these two additional parameters are intended for, but they should have default values like `renderer`, or `renderer` should be placed at last position in the params list. Failing to do so makes `render_notifications` helper method unusable.

Also these two parameters are not mentioned in README documentation.

**Comment by jonhue** [comment]
@artificis

You're absolutely right. Both parameters should default to `nil`. Maybe we should even consider using an `options` hash.

The parameters are being used to make grouping information available to single notification renderers.

Changing the default value for these parameters will also require some work on the `notification_grouped?` method.

**Comment by artificis** [comment]
Hey, thanks for quick reply.
I'm also wondering why we need a thing like `default_type` in `notification_renderer`. After migrating (i.e., adding `type` field to `notifications` table), `user.notifications` keeps scoping `where(type: 'Notification')` to SQL query and it's really annoying because the only way to remove this is `user.notifications.unscope(where: :type)`, which I don't think is pretty looking code.

**Comment by jonhue** [comment]
@artificis

This behavior actually is intended. I don't see many scenarios in which unscoping would be necessary.
However, I would love to learn more about your implementation. Maybe there is a way to optimize it.

**Comment by artificis** [comment]
Why not using `object_type` instead of another `type` field, to classify notification types?

**Comment by jonhue** [comment]
@artificis

What `object_type` column are you pointing towards? `type` is the only column added by NotificationRenderer.

**Comment by jonhue** [comment]
@artificis 

Ah..., I think you misunderstood that column. It is part of the `object` (polymorphic) association of the `Notification` model and cannot be used for purposes other than storing information about the class of a particular ActiveRecord object.

Learn more about notification objects [here](https://github.com/jonhue/notifications-rails/tree/master/notification-handler#notification_object)

**Comment by artificis** [comment]
I knew it. It's just that `object_type` is just enough in my case, to determine notification type.
But I still don't think it helps to scope notifications AR query by default type.

E.g., I have a notification type called "group_invitation". The configuration set default type to "notification". So whenever I try `user.notifications`, it scopes to `type = 'Notification'` by default. What if I want to render all notification types? Do I need to write `user.notifications.unscope(where: :type).where(type: 'group_invitation')`?

Also, `Notification.notification_type`, `Notification.follow_type` etc - these dynamic methods didn't work for me (just threw undefined error). And I'm a bit confused about notification type value format - is it snake case or camel case? (`notification` v `Notification`, `group_invitation` v `GroupInvitation`)?

**Comment by jonhue** [comment]
@artificis 

I'm sorry I think I misunderstood you. The scoping of the actual association to the default type is not intended. Can you verify that that is what's happening as I am unsure what might be causing this.

Just to clarify things, notification types are not meant to represent classes. Their purpose is to structure your renderers & to allow some level of customizablity and adaptivity of your renderers to also work well in large and complex applications.

Types could be used for example to use different renderers for invitation notifications and comment notifications. You could create a type `invitation` and a type `comment`.

Regarding `notification_type` and `follow_type`, those are bugs I think they should be fixed on the master branch.

Notification types can be any string you like. I would suggest snake case, but that's a matter of personal preference.

To keep the conversation organized please create a new issue and link to this discussion.

**Comment by jonhue** [comment]
@artificis 

I did some testing and I cannot reproduce this. Here is the SQL query I get when using the `notifications` association:

```
$ a.notifications
  Notification Load (0.0ms)  SELECT `notifications`.* FROM `notifications` WHERE `notifications`.`type` IN ('Notification') AND `notifications`.`target_id` = 3 AND `notifications`.`target_type` = 'Author'
=> []
```

**Comment by artificis** [comment]
Are you sure you cannot see the issue in above query? :)

> WHERE `notifications`.`type` IN ('Notification') AND ...

**Comment by jonhue** [comment]
@artificis

Whops. Overlooked that. Do you have any idea what might be causing this?
The association looks normal to me: https://github.com/jonhue/notifications-rails/blob/master/notification-handler/lib/notification_handler/target.rb#L10

**Comment by artificis** [comment]
Yes, that's exactly what I thought was really weird. I looked through the whole 4 notifications-rails subgems, tried to find a line that makes that scope, but couldn't find out.

**Comment by artificis** [comment]
Really strange behavior that the scope is made up even with NotificationHandler only (without other 3 gems), as long as `type` field is there in `notifications` table.

**Comment by jonhue** [comment]
@artificis

Ah that makes sense. I believe to remember that Rails has some default behavior for the `type` attribute. Active record uses it for STI.
Either we rename the column or we set a custom inheritance column for the `Notification` model (https://stackoverflow.com/questions/11676288/prevent-sti-when-inheriting-from-an-activerecord-model). I think the latter would be better. What you you propose?

**Comment by jonhue** [comment]
@artificis 

Published 1.2.3

**Comment by jonhue** [comment]
`Notification.notification_type` and `Notification.follow_type` do work now as well.
